Content.Shared/TapeRecorder/SharedTapeRecorderSystem.cs

```diff
@@ -0,0 +1,610 @@
+    [Dependency] private readonly ItemSlotsSystem _itemSlotsSystem = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+    [Dependency] private readonly TagSystem _tagSystem = default!;
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly SharedUserInterfaceSystem _uiSystem = default!;
+    [Dependency] protected readonly IGameTiming Timing = default!; 
+
+    protected const string SlotName = "cassette_tape";
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<TapeRecorderComponent, ItemSlotEjectAttemptEvent>(OnCassetteRemoveAttempt);
+        SubscribeLocalEvent<TapeRecorderComponent, EntRemovedFromContainerMessage>(OnCassetteRemoved);
+        SubscribeLocalEvent<TapeRecorderComponent, EntInsertedIntoContainerMessage>(OnCassetteInserted);
+        SubscribeLocalEvent<TapeRecorderComponent, ExaminedEvent>(OnRecorderExamined);
+        SubscribeLocalEvent<TapeRecorderComponent, ToggleTapeRecorderMessage>(OnToggleMessage);
+        SubscribeLocalEvent<TapeRecorderComponent, ChangeModeTapeRecorderMessage>(OnChangeModeMessage);
+        SubscribeLocalEvent<TapeRecorderComponent, ActivateInWorldEvent>(OnActivate);
+
+        SubscribeLocalEvent<TapeCassetteComponent, ExaminedEvent>(OnTapeExamined);
+        SubscribeLocalEvent<TapeCassetteComponent, DamageChangedEvent>(OnDamagedChanged);
+        SubscribeLocalEvent<TapeCassetteComponent, InteractUsingEvent>(OnInteractingWithCassette);
+        SubscribeLocalEvent<TapeCassetteComponent, TapeCassetteRepairDoAfterEvent>(OnTapeCassetteRepair);
+    }
+
+    /// <summary>
+    /// Process active tape recorder modes
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var tapeRecorders = EntityQueryEnumerator<TapeRecorderComponent>();
+        while (tapeRecorders.MoveNext(out var uid, out var tapeRecorderComponent))
+        {
+            if (tapeRecorderComponent.NeedUIUpdate)
+                UpdateUI((uid, tapeRecorderComponent));
+
+            if (!tapeRecorderComponent.Active)
+                continue;
+
+            switch (tapeRecorderComponent.Mode)
+            {
+                case TapeRecorderMode.Recording:
+                    {
+                        ProcessRecordingTapeRecorder((uid, tapeRecorderComponent), frameTime);
+                        break;
+                    }
+                case TapeRecorderMode.Playing:
+                    {
+                        ProcessPlayingTapeRecorder((uid, tapeRecorderComponent), frameTime);
+                        break;
+                    }
+                case TapeRecorderMode.Rewinding:
+                    {
+                        ProcessRewindingTapeRecorder((uid, tapeRecorderComponent), frameTime);
+                        break;
+                    }
+                default:
+                    continue;
```

Comment:
I don't think this is needed here since it would just reach the end of the while loop and continue anyway?